ra/ra.go

```diff
@@ -1795,8 +1819,327 @@ func (ra *RegistrationAuthorityImpl) revokeCertificate(ctx context.Context, cert
+	}()
+
+	metadata, err := ra.SA.GetSerialMetadata(ctx, &sapb.Serial{Serial: serialString})
+	if err != nil {
+		return nil, err
+	}
+
+	if req.RegID == metadata.RegistrationID {
+		// The requester is the original subscriber. They can revoke for any reason.
+		logEvent.Method = "subscriber"
+	} else {
+		// The requester is a different account. We need to confirm that they have
+		// authorizations for all names in the cert.
+		logEvent.Method = "control"
+
+		authzMapPB, err := ra.SA.GetValidAuthorizations2(ctx, &sapb.GetValidAuthorizationsRequest{
+			RegistrationID: req.RegID,
+			Domains:        cert.DNSNames,
+			Now:            ra.clk.Now().UnixNano(),
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		m := make(map[string]struct{})
+		for _, authz := range authzMapPB.Authz {
+			m[authz.Domain] = struct{}{}
+		}
+		for _, name := range cert.DNSNames {
+			if _, present := m[name]; !present {
+				return nil, berrors.UnauthorizedError("requester does not control all names in cert with serial %q", serialString)
+			}
+		}
+
+		if features.Enabled(features.MozRevocationReasons) {
+			// Applicants who are not the original Subscriber are not allowed to
+			// revoke for any reason other than cessationOfOperation, which covers
+			// circumstances where "the certificate subscriber no longer owns the
+			// domain names in the certificate". Override the reason code to match.
+			req.Code = ocsp.CessationOfOperation
+			logEvent.Reason = req.Code
+		}
+	}
+
+	issuerID := issuance.GetIssuerNameID(cert)
+	err = ra.revokeCertificate(
+		ctx,
+		cert.SerialNumber,
+		int64(issuerID),
+		revocation.Reason(req.Code),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(#5979): Check this error when it can't simply be due to a full queue.
+	_ = ra.purgeOCSPCache(ctx, cert, int64(issuerID))
+
+	return &emptypb.Empty{}, nil
+}
+
+// RevokeCertByKey revokes the certificate in question. It always uses
+// reason code 1 (keyCompromise). It ensures that they public key is added to
+// the blocked keys list, even if revocation otherwise fails. It attempts to
+// purge the certificate from the Akamai cache, but it does not hard-fail if
+// doing so is not successful, because the cache will drop the old OCSP response
+// in less than 24 hours anyway.
+func (ra *RegistrationAuthorityImpl) RevokeCertByKey(ctx context.Context, req *rapb.RevokeCertByKeyRequest) (*emptypb.Empty, error) {
+	if req == nil || req.Cert == nil {
+		return nil, errIncompleteGRPCRequest
+	}
+
+	var reason int64
+	if features.Enabled(features.MozRevocationReasons) {
+		// Upcoming Mozilla policy may require that a certificate be revoked with
+		// reason keyCompromise if "the CA obtains verifiable evidence that the
+		// certificate subscriber’s private key corresponding to the public key in
+		// the certificate suffered a key compromise". Signing a JWS to an ACME
+		// server's revocation endpoint certainly counts, so override the reason.
```

Comment:
"Signing a JWS to an ACME server's revocation endpoint counts" -- if the ACME request said "keyCompromise". However, if the subscriber signs their own request and specifies some other revocation reason, the act of signing the JWS doesn't actually compromise the key.

In other words, I don't think the upcoming Mozilla policy requires us to always set keyCompromise on revoke-by-key requests if the ACME request specified some other code.

Reply:
Ah yes, based on the language you just quoted I believe we have to implement this way. Can you update the comment to include that?